src/world/tree_generation.js

```diff
@@ -1,20 +1,112 @@
 import { alea } from "../libs/alea/alea.js";
 import { rangeFrom, rangeList } from "../utils.js";
-import { SeedFork } from "./seed.js";
+import { BaseGenerator } from "./base_generator.js";
 
 
+export class TreePlacer extends BaseGenerator {
+  constructor(game) {
+    super(game);
+  }
+
+  getGenerator() {
+    switch (this.gcnf.treeCollideAction) {
+      case "avoid":
+        return new AvoidTreePlacer(this);
+      case "place":
+        return new IgnoreTreePlacer(this);
+      case "skip":
+        return new SkipTreePlacer(this);
+      default:
+        throw new ReferenceError("Unknown treePlaceMode");
+    }
+  }
+
+  makeTrees() {
+    return this.getGenerator().makeTrees();
+  }
+}
+
+
+export class IgnoreTreePlacer extends BaseGenerator {
+  constructor(game) {
+    super(game);
+    this.n = this.wSize[0] * this.wSize[2];
+    this.seed = this.getSeed("tree-pos", 0);
+    this.rng = alea(this.seed);
+  }
+
+  makeTrees() {
+    return rangeList(this.n).map(_ => this.idxToCoord(this.rng.randint(0, this.n)));
+  }
+  /**
+   * @param {number} idx
+   * @retuns {[number, number]}
+   */
+  idxToCoord(idx) {
+    let z = Math.floor(idx / this.wSize[0]);
+    let x = idx % this.wSize[0];
+    return [x, z];
+  }
+}
+
+
+export class SkipTreePlacer extends BaseGenerator {
+  constructor(game) {
+    super(game);
+    this.n = this.wSize[0] * this.wSize[2];
+    this.seed = this.getSeed("tree-pos", 0);
+    this.rng = alea(this.seed);
+    let w = this.gcnf.treeRadius[0];
```

Comment:
There is a lot of common functionality & duplicated code between the 3 tree placers. Perhaps make a `BaseTreeGenerator` class have have them inherit from it.